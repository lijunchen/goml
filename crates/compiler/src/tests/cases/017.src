trait ToString {
    fn to_string(Self) -> String;
}

impl ToString for Int {
    fn to_string(self: Int) -> String {
        int_to_string(self)
    }
}

impl ToString for Bool  {
    fn to_string(self: Bool) -> String {
        bool_to_string(self)
    }
}

impl ToString for (Int, Int) {
  fn to_string(self: (Int, Int)) -> String {
    "(?, ?)"
  }
}

fn main() {
    let x = 123 in
    let _ string_println(to_string(x)) in

    let x = true in
    let _ = string_println(to_string(x)) in

    let x = (3, 4) in
    let _ = string_println(to_string(x)) in

    ()
}
