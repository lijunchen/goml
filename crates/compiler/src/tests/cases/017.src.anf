fn impl_ToString_int_to_string(self/0: int) -> string {
  int_to_string(self/0)
}

fn impl_ToString_bool_to_string(self/1: bool) -> string {
  bool_to_string(self/1)
}

fn impl_ToString_Tuple_int_int_to_string(self/2: (int, int)) -> string {
  "(?, ?)"
}

fn main() -> unit {
  let x/3 = 123 in
  let t3 = impl_ToString_int_to_string(x/3) in
  let mtmp0 = string_println(t3) in
  let x/4 = true in
  let t4 = impl_ToString_bool_to_string(x/4) in
  let mtmp1 = string_println(t4) in
  let x/5 = (3, 4) in
  let t5 = impl_ToString_Tuple_int_int_to_string(x/5) in
  let mtmp2 = string_println(t5) in
  ()
}