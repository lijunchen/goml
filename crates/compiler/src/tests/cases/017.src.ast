trait ToString {
  fn to_string(Self) -> String;
}

impl ToString for Int {
  fn to_string(self: Int) -> String {
      int_to_string(self)
  }
}

impl ToString for Bool {
  fn to_string(self: Bool) -> String {
      bool_to_string(self)
  }
}

fn main() {
    let x = 123 in
    let _ = println(to_string(x)) in
    let x = true in
    let _ = println(to_string(x)) in
    ()
}

