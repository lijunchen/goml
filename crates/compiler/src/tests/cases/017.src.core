fn impl_ToString_Int_to_string(self/0: Int) -> String {
  int_to_string(self/0)
}

fn impl_ToString_Bool_to_string(self/1: Bool) -> String {
  bool_to_string(self/1)
}

fn main() -> Unit {
  let x/2 = 123 in
  let mtmp0 = println(impl_ToString_Int_to_string(x/2)) in
  let x/3 = true in
  let mtmp1 = println(impl_ToString_Bool_to_string(x/3)) in
  ()
}