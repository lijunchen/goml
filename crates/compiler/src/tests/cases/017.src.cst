FILE@0..573
  TRAIT@0..54
    TraitKeyword@0..5 "trait"
    Whitespace@5..6 " "
    Uident@6..14 "ToString"
    Whitespace@14..15 " "
    LBrace@15..16 "{"
    Whitespace@16..21 "\n    "
    TRAIT_METHOD_SIG_LIST@21..54
      TRAIT_METHOD_SIG@21..49
        FnKeyword@21..23 "fn"
        Whitespace@23..24 " "
        Lident@24..33 "to_string"
        TYPE_LIST@33..40
          LParen@33..34 "("
          TYPE_TAPP@34..38
            Uident@34..38 "Self"
          RParen@38..39 ")"
          Whitespace@39..40 " "
        Arrow@40..42 "->"
        Whitespace@42..43 " "
        TYPE_STRING@43..49
          StringKeyword@43..49 "string"
      Semi@49..50 ";"
      Whitespace@50..51 "\n"
      RBrace@51..52 "}"
      Whitespace@52..54 "\n\n"
  IMPL@54..155
    ImplKeyword@54..58 "impl"
    Whitespace@58..59 " "
    Uident@59..67 "ToString"
    Whitespace@67..68 " "
    ForKeyword@68..71 "for"
    Whitespace@71..72 " "
    TYPE_INT@72..76
      IntKeyword@72..75 "int"
      Whitespace@75..76 " "
    LBrace@76..77 "{"
    Whitespace@77..82 "\n    "
    FN@82..152
      FnKeyword@82..84 "fn"
      Whitespace@84..85 " "
      Lident@85..94 "to_string"
      PARAM_LIST@94..106
        LParen@94..95 "("
        PARAM@95..104
          Lident@95..99 "self"
          Colon@99..100 ":"
          Whitespace@100..101 " "
          TYPE_INT@101..104
            IntKeyword@101..104 "int"
        RParen@104..105 ")"
        Whitespace@105..106 " "
      Arrow@106..108 "->"
      Whitespace@108..109 " "
      TYPE_STRING@109..116
        StringKeyword@109..115 "string"
        Whitespace@115..116 " "
      BLOCK@116..152
        LBrace@116..117 "{"
        Whitespace@117..126 "\n        "
        EXPR_CALL@126..150
          EXPR_LIDENT@126..139
            Lident@126..139 "int_to_string"
          ARG_LIST@139..150
            LParen@139..140 "("
            ARG@140..144
              EXPR_LIDENT@140..144
                Lident@140..144 "self"
            RParen@144..145 ")"
            Whitespace@145..150 "\n    "
        RBrace@150..151 "}"
        Whitespace@151..152 "\n"
    RBrace@152..153 "}"
    Whitespace@153..155 "\n\n"
  IMPL@155..260
    ImplKeyword@155..159 "impl"
    Whitespace@159..160 " "
    Uident@160..168 "ToString"
    Whitespace@168..169 " "
    ForKeyword@169..172 "for"
    Whitespace@172..173 " "
    TYPE_BOOL@173..179
      BoolKeyword@173..177 "bool"
      Whitespace@177..179 "  "
    LBrace@179..180 "{"
    Whitespace@180..185 "\n    "
    FN@185..257
      FnKeyword@185..187 "fn"
      Whitespace@187..188 " "
      Lident@188..197 "to_string"
      PARAM_LIST@197..210
        LParen@197..198 "("
        PARAM@198..208
          Lident@198..202 "self"
          Colon@202..203 ":"
          Whitespace@203..204 " "
          TYPE_BOOL@204..208
            BoolKeyword@204..208 "bool"
        RParen@208..209 ")"
        Whitespace@209..210 " "
      Arrow@210..212 "->"
      Whitespace@212..213 " "
      TYPE_STRING@213..220
        StringKeyword@213..219 "string"
        Whitespace@219..220 " "
      BLOCK@220..257
        LBrace@220..221 "{"
        Whitespace@221..230 "\n        "
        EXPR_CALL@230..255
          EXPR_LIDENT@230..244
            Lident@230..244 "bool_to_string"
          ARG_LIST@244..255
            LParen@244..245 "("
            ARG@245..249
              EXPR_LIDENT@245..249
                Lident@245..249 "self"
            RParen@249..250 ")"
            Whitespace@250..255 "\n    "
        RBrace@255..256 "}"
        Whitespace@256..257 "\n"
    RBrace@257..258 "}"
    Whitespace@258..260 "\n\n"
  IMPL@260..356
    ImplKeyword@260..264 "impl"
    Whitespace@264..265 " "
    Uident@265..273 "ToString"
    Whitespace@273..274 " "
    ForKeyword@274..277 "for"
    Whitespace@277..278 " "
    TYPE_TUPLE@278..289
      TYPE_LIST@278..289
        LParen@278..279 "("
        TYPE_INT@279..282
          IntKeyword@279..282 "int"
        Comma@282..283 ","
        Whitespace@283..284 " "
        TYPE_INT@284..287
          IntKeyword@284..287 "int"
        RParen@287..288 ")"
        Whitespace@288..289 " "
    LBrace@289..290 "{"
    Whitespace@290..293 "\n  "
    FN@293..353
      FnKeyword@293..295 "fn"
      Whitespace@295..296 " "
      Lident@296..305 "to_string"
      PARAM_LIST@305..324
        LParen@305..306 "("
        PARAM@306..322
          Lident@306..310 "self"
          Colon@310..311 ":"
          Whitespace@311..312 " "
          TYPE_TUPLE@312..322
            TYPE_LIST@312..322
              LParen@312..313 "("
              TYPE_INT@313..316
                IntKeyword@313..316 "int"
              Comma@316..317 ","
              Whitespace@317..318 " "
              TYPE_INT@318..321
                IntKeyword@318..321 "int"
              RParen@321..322 ")"
        RParen@322..323 ")"
        Whitespace@323..324 " "
      Arrow@324..326 "->"
      Whitespace@326..327 " "
      TYPE_STRING@327..334
        StringKeyword@327..333 "string"
        Whitespace@333..334 " "
      BLOCK@334..353
        LBrace@334..335 "{"
        Whitespace@335..340 "\n    "
        EXPR_STR@340..351
          Str@340..348 "\"(?, ?)\""
          Whitespace@348..351 "\n  "
        RBrace@351..352 "}"
        Whitespace@352..353 "\n"
    RBrace@353..354 "}"
    Whitespace@354..356 "\n\n"
  FN@356..573
    FnKeyword@356..358 "fn"
    Whitespace@358..359 " "
    Lident@359..363 "main"
    PARAM_LIST@363..366
      LParen@363..364 "("
      RParen@364..365 ")"
      Whitespace@365..366 " "
    BLOCK@366..573
      LBrace@366..367 "{"
      Whitespace@367..372 "\n    "
      EXPR_LET@372..571
        LetKeyword@372..375 "let"
        Whitespace@375..376 " "
        PATTERN_VARIABLE@376..378
          Lident@376..377 "x"
          Whitespace@377..378 " "
        Eq@378..379 "="
        Whitespace@379..380 " "
        EXPR_LET_VALUE@380..384
          EXPR_INT@380..384
            Int@380..383 "123"
            Whitespace@383..384 " "
        InKeyword@384..386 "in"
        Whitespace@386..391 "\n    "
        EXPR_LET_BODY@391..571
          EXPR_LET@391..571
            LetKeyword@391..394 "let"
            Whitespace@394..395 " "
            PATTERN_WILDCARD@395..397
              WildcardKeyword@395..396 "_"
              Whitespace@396..397 " "
            Eq@397..398 "="
            Whitespace@398..399 " "
            EXPR_LET_VALUE@399..428
              EXPR_CALL@399..428
                EXPR_LIDENT@399..413
                  Lident@399..413 "string_println"
                ARG_LIST@413..428
                  LParen@413..414 "("
                  ARG@414..426
                    EXPR_CALL@414..426
                      EXPR_LIDENT@414..423
                        Lident@414..423 "to_string"
                      ARG_LIST@423..426
                        LParen@423..424 "("
                        ARG@424..425
                          EXPR_LIDENT@424..425
                            Lident@424..425 "x"
                        RParen@425..426 ")"
                  RParen@426..427 ")"
                  Whitespace@427..428 " "
            InKeyword@428..430 "in"
            Whitespace@430..436 "\n\n    "
            EXPR_LET_BODY@436..571
              EXPR_LET@436..571
                LetKeyword@436..439 "let"
                Whitespace@439..440 " "
                PATTERN_VARIABLE@440..442
                  Lident@440..441 "x"
                  Whitespace@441..442 " "
                Eq@442..443 "="
                Whitespace@443..444 " "
                EXPR_LET_VALUE@444..449
                  EXPR_BOOL@444..449
                    TrueKeyword@444..448 "true"
                    Whitespace@448..449 " "
                InKeyword@449..451 "in"
                Whitespace@451..456 "\n    "
                EXPR_LET_BODY@456..571
                  EXPR_LET@456..571
                    LetKeyword@456..459 "let"
                    Whitespace@459..460 " "
                    PATTERN_WILDCARD@460..462
                      WildcardKeyword@460..461 "_"
                      Whitespace@461..462 " "
                    Eq@462..463 "="
                    Whitespace@463..464 " "
                    EXPR_LET_VALUE@464..493
                      EXPR_CALL@464..493
                        EXPR_LIDENT@464..478
                          Lident@464..478 "string_println"
                        ARG_LIST@478..493
                          LParen@478..479 "("
                          ARG@479..491
                            EXPR_CALL@479..491
                              EXPR_LIDENT@479..488
                                Lident@479..488 "to_string"
                              ARG_LIST@488..491
                                LParen@488..489 "("
                                ARG@489..490
                                  EXPR_LIDENT@489..490
                                    Lident@489..490 "x"
                                RParen@490..491 ")"
                          RParen@491..492 ")"
                          Whitespace@492..493 " "
                    InKeyword@493..495 "in"
                    Whitespace@495..501 "\n\n    "
                    EXPR_LET_BODY@501..571
                      EXPR_LET@501..571
                        LetKeyword@501..504 "let"
                        Whitespace@504..505 " "
                        PATTERN_VARIABLE@505..507
                          Lident@505..506 "x"
                          Whitespace@506..507 " "
                        Eq@507..508 "="
                        Whitespace@508..509 " "
                        EXPR_LET_VALUE@509..516
                          EXPR_TUPLE@509..516
                            LParen@509..510 "("
                            EXPR_INT@510..511
                              Int@510..511 "3"
                            Comma@511..512 ","
                            Whitespace@512..513 " "
                            EXPR_INT@513..514
                              Int@513..514 "4"
                            RParen@514..515 ")"
                            Whitespace@515..516 " "
                        InKeyword@516..518 "in"
                        Whitespace@518..523 "\n    "
                        EXPR_LET_BODY@523..571
                          EXPR_LET@523..571
                            LetKeyword@523..526 "let"
                            Whitespace@526..527 " "
                            PATTERN_WILDCARD@527..529
                              WildcardKeyword@527..528 "_"
                              Whitespace@528..529 " "
                            Eq@529..530 "="
                            Whitespace@530..531 " "
                            EXPR_LET_VALUE@531..560
                              EXPR_CALL@531..560
                                EXPR_LIDENT@531..545
                                  Lident@531..545 "string_println"
                                ARG_LIST@545..560
                                  LParen@545..546 "("
                                  ARG@546..558
                                    EXPR_CALL@546..558
                                      EXPR_LIDENT@546..555
                                        Lident@546..555 "to_string"
                                      ARG_LIST@555..558
                                        LParen@555..556 "("
                                        ARG@556..557
                                          EXPR_LIDENT@556..557
                                            Lident@556..557 "x"
                                        RParen@557..558 ")"
                                  RParen@558..559 ")"
                                  Whitespace@559..560 " "
                            InKeyword@560..562 "in"
                            Whitespace@562..568 "\n\n    "
                            EXPR_LET_BODY@568..571
                              EXPR_UNIT@568..571
                                LParen@568..569 "("
                                RParen@569..570 ")"
                                Whitespace@570..571 "\n"
      RBrace@571..572 "}"
      Whitespace@572..573 "\n"