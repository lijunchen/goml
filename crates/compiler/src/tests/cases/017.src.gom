package main

import (
    "fmt"
)

func unit_to_string(x struct{}) string {
    return "()"
}

func bool_to_string(x bool) string {
    if x {
        return "true"
    } else {
        return "false"
    }
}

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func int_add(x int, y int) int {
    return x + y
}

func int_sub(x int, y int) int {
    return x - y
}

func int_less(x int, y int) bool {
    return x < y
}

func string_add(a string, b string) string {
    return a + b
}

func string_print(s string) struct{} {
    fmt.Print(s)
    return struct{}{}
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

func missing(s string) struct{} {
    println("missing: " + s)
    panic("")
    return struct{}{}
}

type Tuple2_int_int struct {
    _0 int
    _1 int
}

func impl_ToString_int_to_string(self__0 int) string {
    var ret6 string
    ret6 = int_to_string(self__0)
    return ret6
}

func impl_ToString_bool_to_string(self__1 bool) string {
    var ret7 string
    ret7 = bool_to_string(self__1)
    return ret7
}

func impl_ToString_Tuple_TInt_TInt_to_string(self__2 Tuple2_int_int) string {
    var ret8 string
    ret8 = "(?, ?)"
    return ret8
}

func main0() struct{} {
    var ret9 struct{}
    var x__3 int = 123
    var t3 string = impl_ToString_int_to_string(x__3)
    string_println(t3)
    var x__4 bool = true
    var t4 string = impl_ToString_bool_to_string(x__4)
    string_println(t4)
    var x__5 Tuple2_int_int = Tuple2_int_int{
        _0: 3,
        _1: 4,
    }
    var t5 string = impl_ToString_Tuple_TInt_TInt_to_string(x__5)
    string_println(t5)
    ret9 = struct{}{}
    return ret9
}

func main() {
    main0()
}
