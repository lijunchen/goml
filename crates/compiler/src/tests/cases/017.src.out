()
====
123
true
(?, ?)

====
123
true
(?, ?)
